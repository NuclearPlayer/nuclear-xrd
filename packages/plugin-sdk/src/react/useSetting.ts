import { useEffect, useMemo, useRef, useState } from 'react';

import type { SettingsHost, SettingValue } from '../types/settings';

export const useSetting = <T extends SettingValue = SettingValue>(
  host: SettingsHost | undefined,
  id: string,
) => {
  const hostRef = useRef(host);
  hostRef.current = host;

  const [currentValue, setCurrentValue] = useState<T | undefined>(undefined);

  useEffect(() => {
    if (!hostRef.current) return;
    let unsubscribe: (() => void) | undefined;
    let isMounted = true;

    hostRef.current.get<T>(id).then((initialValue) => {
      if (!isMounted) return;
      setCurrentValue(initialValue);
      unsubscribe = hostRef.current?.subscribe<T>(id, (nextValue) =>
        setCurrentValue(nextValue),
      );
    });

    return () => {
      isMounted = false;
      if (unsubscribe) unsubscribe();
    };
  }, [id, hostRef]);

  const setValue = useMemo(
    () => (nextValue: T) => {
      if (!hostRef.current) return;
      void hostRef.current.set<T>(id, nextValue);
    },
    [id],
  );

  return [currentValue, setValue] as const;
};
