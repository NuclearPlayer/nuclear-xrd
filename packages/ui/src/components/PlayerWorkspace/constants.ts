export const SIDEBAR_CONFIG = {
  COLLAPSED_WIDTH: 56,
  MIN_WIDTH: 150,
  MAX_WIDTH: 400,
  RESIZE_HANDLE_WIDTH: 1,
  TOGGLE_BUTTON_SIZE: 4,
  TOGGLE_BUTTON_OFFSET: 2,
} as const;

export const SIDEBAR_ANIMATIONS = {
  width: {
    spring: {
      type: 'spring' as const,
      stiffness: 300,
      damping: 30,
      mass: 0.8,
    },
    instant: { duration: 0 },
  },
  content: {
    spring: {
      type: 'spring' as const,
      stiffness: 400,
      damping: 25,
      mass: 0.5,
    },
    slideDistance: 20,
  },
  button: {
    hover: { scale: 1.1 },
    tap: { scale: 0.95 },
    spring: {
      type: 'spring' as const,
      stiffness: 400,
      damping: 17,
    },
    rotation: {
      type: 'spring' as const,
      stiffness: 300,
      damping: 20,
    },
  },
} as const;
