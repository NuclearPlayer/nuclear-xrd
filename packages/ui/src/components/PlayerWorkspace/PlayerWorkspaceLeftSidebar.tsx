import { FC } from 'react';

import {
  PlayerWorkspaceSidebar,
  PlayerWorkspaceSidebarPropsBase,
} from './PlayerWorkspaceSidebar';

export const PlayerWorkspaceLeftSidebar: FC<
  PlayerWorkspaceSidebarPropsBase
> = ({ children, ...props }) => {
  return (
    <PlayerWorkspaceSidebar side="left" {...props}>
      {children}
    </PlayerWorkspaceSidebar>
  );
};
