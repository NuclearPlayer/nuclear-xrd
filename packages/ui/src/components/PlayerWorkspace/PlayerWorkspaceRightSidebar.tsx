import { FC } from 'react';

import {
  PlayerWorkspaceSidebar,
  PlayerWorkspaceSidebarPropsBase,
} from './PlayerWorkspaceSidebar';

export const PlayerWorkspaceRightSidebar: FC<
  PlayerWorkspaceSidebarPropsBase
> = ({ children, ...props }) => {
  return (
    <PlayerWorkspaceSidebar side="right" {...props}>
      {children}
    </PlayerWorkspaceSidebar>
  );
};
