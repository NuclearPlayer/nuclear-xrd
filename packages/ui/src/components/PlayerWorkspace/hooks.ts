import { useCallback, useRef, useState } from 'react';

import { SIDEBAR_CONFIG } from './constants';

export const useSidebarResize = (
  width: number,
  onWidthChange: (width: number) => void,
  side: 'left' | 'right',
  isCollapsed: boolean,
) => {
  const isResizing = useRef(false);
  const [isResizingState, setIsResizingState] = useState(false);

  const handleMouseDown = useCallback(
    (e: React.MouseEvent) => {
      if (isCollapsed) return;

      isResizing.current = true;
      setIsResizingState(true);
      e.preventDefault();

      const startX = e.clientX;
      const startWidth = width;

      const handleMouseMove = (e: MouseEvent) => {
        if (!isResizing.current) return;

        const deltaX =
          side === 'left' ? e.clientX - startX : startX - e.clientX;
        const newWidth = Math.max(
          SIDEBAR_CONFIG.MIN_WIDTH,
          Math.min(SIDEBAR_CONFIG.MAX_WIDTH, startWidth + deltaX),
        );
        onWidthChange(newWidth);
      };

      const handleMouseUp = () => {
        isResizing.current = false;
        setIsResizingState(false);
        document.removeEventListener('mousemove', handleMouseMove);
        document.removeEventListener('mouseup', handleMouseUp);
      };

      document.addEventListener('mousemove', handleMouseMove);
      document.addEventListener('mouseup', handleMouseUp);
    },
    [width, onWidthChange, side, isCollapsed],
  );

  return { handleMouseDown, isResizingState };
};
