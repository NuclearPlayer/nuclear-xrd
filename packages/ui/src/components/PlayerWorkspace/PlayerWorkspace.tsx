import { FC, ReactNode } from 'react';

import { cn } from '../../utils';
import { PlayerWorkspaceLeftSidebar } from './PlayerWorkspaceLeftSidebar';
import { PlayerWorkspaceRightSidebar } from './PlayerWorkspaceRightSidebar';

type PlayerWorkspaceProps = {
  children: ReactNode;
  className?: string;
};

type MainProps = {
  children?: ReactNode;
  className?: string;
};

const PlayerWorkspaceMain: FC<MainProps> = ({ children, className = '' }) => {
  return (
    <main
      data-testid="player-workspace-main"
      className={cn('bg-background-secondary overflow-auto', className)}
    >
      {children}
    </main>
  );
};

type PlayerWorkspaceComponent = FC<PlayerWorkspaceProps> & {
  LeftSidebar: typeof PlayerWorkspaceLeftSidebar;
  RightSidebar: typeof PlayerWorkspaceRightSidebar;
  Main: typeof PlayerWorkspaceMain;
};

export const PlayerWorkspace: PlayerWorkspaceComponent = Object.assign(
  ({ children, className = '' }: PlayerWorkspaceProps) => {
    return (
      <div
        className={cn(
          'grid h-full min-h-0 grid-cols-[auto_1fr_auto] bg-background-secondary relative',
          className,
        )}
      >
        {children}
      </div>
    );
  },
  {
    LeftSidebar: PlayerWorkspaceLeftSidebar,
    RightSidebar: PlayerWorkspaceRightSidebar,
    Main: PlayerWorkspaceMain,
  },
);
