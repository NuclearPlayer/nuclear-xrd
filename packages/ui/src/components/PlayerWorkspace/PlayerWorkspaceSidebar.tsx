import clsx from 'clsx';
import { AnimatePresence, motion } from 'framer-motion';
import { PanelLeft, PanelRight } from 'lucide-react';
import { FC, ReactNode, useRef } from 'react';

import { Button } from '../Button';
import { SIDEBAR_CONFIG } from './constants';
import { useSidebarResize } from './hooks';

export type PlayerWorkspaceSidebarPropsBase = {
  children?: ReactNode;
  isCollapsed: boolean;
  width: number;
  onWidthChange: (width: number) => void;
  onToggle: () => void;
  className?: string;
};

type PlayerWorkspaceSidebarProps = PlayerWorkspaceSidebarPropsBase & {
  side: 'left' | 'right';
};

export const PlayerWorkspaceSidebar: FC<PlayerWorkspaceSidebarProps> = ({
  children,
  isCollapsed,
  width,
  onWidthChange,
  onToggle,
  side,
  className = '',
}) => {
  const sidebarRef = useRef<HTMLDivElement>(null);
  const { handleMouseDown, isResizingState } = useSidebarResize(
    width,
    onWidthChange,
    side,
    isCollapsed,
  );

  const currentWidth = isCollapsed ? SIDEBAR_CONFIG.COLLAPSED_WIDTH : width;

  return (
    <motion.div
      ref={sidebarRef}
      className={clsx(
        'bg-background-secondary border-border relative flex flex-col p-2',
        { 'border-r-2': side === 'left', 'border-l-2': side === 'right' },
        className,
      )}
      animate={{ width: currentWidth }}
      transition={
        isResizingState
          ? { duration: 0 }
          : {
              type: 'spring',
              stiffness: 300,
              damping: 30,
              mass: 0.8,
            }
      }
    >
      <span
        className={clsx('mb-4 flex flex-row items-center', {
          'justify-end': side === 'left',
          'justify-start': side === 'right',
        })}
      >
        <Button
          data-testid={`sidebar-toggle-${side}`}
          className={clsx('top-2 px-2', {
            'right-1': side === 'left',
            'left-1': side === 'right',
          })}
          size="icon"
          onClick={onToggle}
        >
          {side === 'left' ? <PanelLeft /> : <PanelRight />}
        </Button>
      </span>
      <AnimatePresence mode="wait">
        {!isCollapsed && (
          <motion.div
            className="flex-1 overflow-hidden"
            initial={{ opacity: 0, x: side === 'left' ? -20 : 20 }}
            animate={{ opacity: 1, x: 0 }}
            exit={{ opacity: 0, x: side === 'left' ? -20 : 20 }}
            transition={{
              type: 'spring',
              stiffness: 400,
              damping: 25,
              mass: 0.5,
            }}
          >
            {children}
          </motion.div>
        )}
      </AnimatePresence>

      {!isCollapsed && (
        <div
          className={`absolute top-0 bottom-0 w-1 cursor-col-resize transition-colors ${
            side === 'left' ? 'right-0' : 'left-0'
          }`}
          onMouseDown={handleMouseDown}
        />
      )}
    </motion.div>
  );
};
