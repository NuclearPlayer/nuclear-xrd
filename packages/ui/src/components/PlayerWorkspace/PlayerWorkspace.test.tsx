import { render } from '@testing-library/react';

import { PlayerWorkspace } from './PlayerWorkspace';

describe('PlayerWorkspace', () => {
  it('(Snapshot) both sidebars expanded', () => {
    const { asFragment } = render(
      <PlayerWorkspace>
        <PlayerWorkspace.LeftSidebar
          width={200}
          isCollapsed={false}
          onWidthChange={() => {}}
          onToggle={() => {}}
        />
        <PlayerWorkspace.Main />
        <PlayerWorkspace.RightSidebar
          width={200}
          isCollapsed={false}
          onWidthChange={() => {}}
          onToggle={() => {}}
        />
      </PlayerWorkspace>,
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('(Snapshot) both sidebars collapsed', () => {
    const { asFragment } = render(
      <PlayerWorkspace>
        <PlayerWorkspace.LeftSidebar
          width={200}
          isCollapsed={true}
          onWidthChange={() => {}}
          onToggle={() => {}}
        />
        <PlayerWorkspace.Main />
        <PlayerWorkspace.RightSidebar
          width={200}
          isCollapsed={true}
          onWidthChange={() => {}}
          onToggle={() => {}}
        />
      </PlayerWorkspace>,
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
