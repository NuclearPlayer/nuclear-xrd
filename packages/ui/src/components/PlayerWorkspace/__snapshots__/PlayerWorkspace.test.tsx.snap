// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PlayerWorkspace > (Snapshot) both sidebars collapsed 1`] = `
<DocumentFragment>
  <div
    class="grid grid-cols-[auto_1fr_auto] h-full "
  >
    <div
      class="bg-gray-800 border-gray-700 flex relative border-r "
    >
      <button
        class="absolute top-2 w-4 h-4 bg-gray-600 hover:bg-gray-500 rounded-sm flex items-center justify-center text-xs text-gray-300 right-1"
      >
        ›
      </button>
    </div>
    <main
      class="bg-gray-100 overflow-auto "
    />
    <div
      class="bg-gray-800 border-gray-700 flex relative border-l "
    >
      <button
        class="absolute top-2 w-4 h-4 bg-gray-600 hover:bg-gray-500 rounded-sm flex items-center justify-center text-xs text-gray-300 left-1"
      >
        ‹
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`PlayerWorkspace > (Snapshot) both sidebars expanded 1`] = `
<DocumentFragment>
  <div
    class="grid grid-cols-[auto_1fr_auto] h-full "
  >
    <div
      class="bg-gray-800 border-gray-700 flex relative border-r "
    >
      <div
        class="flex-1 overflow-hidden"
        style="opacity: 0; transform: translateX(-20px);"
      />
      <div
        class="absolute top-0 bottom-0 w-1 cursor-col-resize hover:bg-blue-500 transition-colors right-0"
      />
      <button
        class="absolute top-2 w-4 h-4 bg-gray-600 hover:bg-gray-500 rounded-sm flex items-center justify-center text-xs text-gray-300 right-1"
      >
        ‹
      </button>
    </div>
    <main
      class="bg-gray-100 overflow-auto "
    />
    <div
      class="bg-gray-800 border-gray-700 flex relative border-l "
    >
      <div
        class="flex-1 overflow-hidden"
        style="opacity: 0; transform: translateX(20px);"
      />
      <div
        class="absolute top-0 bottom-0 w-1 cursor-col-resize hover:bg-blue-500 transition-colors left-0"
      />
      <button
        class="absolute top-2 w-4 h-4 bg-gray-600 hover:bg-gray-500 rounded-sm flex items-center justify-center text-xs text-gray-300 left-1"
      >
        ›
      </button>
    </div>
  </div>
</DocumentFragment>
`;
